File: apps/dma-insights/frontend/src/lib/stairstep.ts
/**
 * Stairstep typed hook + types.
 *
 * State-branch contract:
 *   - displayId null            → query disabled, no fetch
 *   - empty_state="no-gaps"     → entity has no scored subcaps yet
 *   - empty_state="no-recs"     → no recommendations to apply
 *   - empty_state="no-applicable-uplift"
 *                               → every pillar already at/above target
 *   - empty_state=null + steps  → happy path
 */
import { useQuery, type UseQueryResult } from "@tanstack/react-query";
import { apiGet } from "./api";

export interface StairStepOut {
  rec_id: string;
  title: string;
  pillar: string;
  score_before: number;
  score_after: number;
  uplift: number;
}

export interface StairstepResponse {
  entity_display_id: string;
  run_request_id: string | null;
  steps_by_pillar: Record<string, StairStepOut[]>;
  current_by_pillar: Record<string, number>;
  end_score_by_pillar: Record<string, number>;
  target_band_score: number;
  empty_state: "no-gaps" | "no-recs" | "no-applicable-uplift" | null;
}

export function useStairstep(
  displayId: string | null,
): UseQueryResult<StairstepResponse> {
  return useQuery({
    queryKey: ["stairstep", displayId],
    queryFn: () => apiGet<StairstepResponse>(`/api/v1/entities/${displayId}/stairstep`),
    enabled: displayId !== null,
    staleTime: 60 * 1000,
  });
}

File: apps/dma-insights/backend/app/schemas/stairstep.py
"""D4 StairstepCurve response schemas."""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class StairStepOut(BaseModel):
    rec_id: str
    title: str
    pillar: str
    score_before: float = Field(..., ge=0.0, le=5.0)
    score_after: float = Field(..., ge=0.0, le=5.0)
    uplift: float = Field(..., ge=0.0)


class StairstepResponse(BaseModel):
    entity_display_id: str
    run_request_id: str | None = None
    steps_by_pillar: dict[str, list[StairStepOut]] = Field(default_factory=dict)
    current_by_pillar: dict[str, float] = Field(default_factory=dict)
    end_score_by_pillar: dict[str, float] = Field(default_factory=dict)
    target_band_score: float = 4.0
    empty_state: Literal["no-gaps", "no-recs", "no-applicable-uplift"] | None = None
    # Names the coarser signal that placed the client on the curve when the
    # run carried no scored subcaps ("pillar_scores" | "overall_maturity"),
    # else None when per-subcap scores drove the position.
    position_source: str | None = None
